moquette-0.16/broker/src/main/java/io/moquette/broker/MQTTConnection.java

```diff
@@ -283,7 +289,21 @@ private boolean login(MqttConnectMessage msg, final String clientId) {
                 LOG.info("Client didn't supply any password and MQTT anonymous mode is disabled CId={}", clientId);
                 return false;
             }
-            final String login = msg.payload().userName();
+            String login = msg.payload().userName();
+            if (brokerConfig.peerCertificateAsUsername()) {
+                try {
+                    // Use peer cert as username
+                    SslHandler sslhandler = (SslHandler) channel.pipeline().get("ssl");
+                    if (sslhandler != null) {
+                        Certificate[] certificateChain = sslhandler.engine().getSession().getPeerCertificates();
+                        login = PemUtils.certificatesToPem(certificateChain);
+                    }
+                } catch (SSLPeerUnverifiedException e) {
+                    // No peer cert provided
+                } catch (CertificateEncodingException|IOException e) {
+                    return false;
+                }
+            }
```

Comment:
This isn't quite right. The check on line 284 means we won't get here if there isn't a username provided.

Interestingly enough, I tested this with the device SDK and it DOES work, which means it is providing a username (although maybe an empty one). Odd.